### PR TITLE
Sherpa romeo integration (Angular)

### LINKED ISSUE #769: Sherpa-Romeo Integration
From release plan spreadsheet

Estimate from release plan: 40 hours
Expressing interest: none

No additional notes

Related REST ticket https://github.com/DSpace/DSpace/issues/2840

### PR DESCRIPTION
**Pull request by atarix83**
## References
_Add references/links to any related issues or PRs. These may include:_
* Fixes #769
* Requires DSpace/DSpace#8271

## Description
This PR implement Sherpa romeo integration during submission form

## Instructions for Reviewers
- Follow instruction in the [rest PR](DSpace/DSpace#8271) for setting up the Sherpa romeo step in the submission configuration.
- Create a new submission and edit the `dc.identifier.issn` (for example use `2731-0582` as id)
- Check the information are shown in the Sherpa romeo section 

List of changes in this PR:
* Add a new submission section for managing the sherpa romeo integration

## Checklist
- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & specs/tests), or I have provided reasons as to why that's not possible.
- [x] My PR passes [TSLint](https://palantir.github.io/tslint/) validation using `yarn run lint`
- [x] My PR doesn't introduce circular dependencies
- [x] My PR includes [TypeDoc](https://typedoc.org/) comments for _all new (or modified) public methods and classes_. It also includes TypeDoc for large or complex private methods.
- [x] My PR passes all specs/tests and includes new/updated specs or tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new, third-party dependencies (in `package.json`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.

### CODE REVIEW

File: src/app/submission/sections/sherpa-policies/content-accordion/content-accordion.component.scss [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+                    margin-right: 10px;
+                    height: 1.25em;
+                    display: flex;
+                    align-items: center;
+                    justify-content: center;
+                    font-size: 12px;
+                }
+            }
+        }
+    }
```
Comment (artlowel) [inline src/app/submission/sections/sherpa-policies/content-accordion/content-accordion.component.scss:31]: I'm not a fan of these big css changes to the default bootstrap components in the base theme. Especially not hardcoded colors, and font-sizes etc, as they will not be influenced by the rest of the theme, and have to be modified separately every time. 

Why not use default bootstap cards, and position the button and icon using utility classes like `bg-light, `text-dark`, etc?

File: src/app/submission/sections/sherpa-policies/section-sherpa-policies.component.scss [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+                    height: 1.25em;
+                    display: flex;
+                    align-items: center;
+                    justify-content: center;
+                }
+            }
+        }
+    }
+
+}
```
Comment (artlowel) [inline src/app/submission/sections/sherpa-policies/section-sherpa-policies.component.scss:31]: Same comment as for ContentAccordionComponent

**Review by artlowel — CHANGES_REQUESTED** [review]
Thanks @atarix83!

**Review by tdonohue — CHANGES_REQUESTED** [review]
@atarix83 : Gave this a test today, and overall it works.  But, I find the layout of the information from Sherpa very confusing. 

* In the UI, it says to "Click on a pathway for a more detailed view", but that's not possible. It looks like all the details are opened automatically.  Are they meant to be closed initially perhaps?  (see screenshot)
* I find the blank entries very distracting. If there's no value in a section we should remove it. (see screenshot)
* The duplicative titles in sections are also odd looking. Perhaps we should mimic Sherpa's UI and next to "Accepted Version" add in parantheses either the *location* or the Embargo, or both?  E.g.
     * "Accepted Version (No Embargo, Locations: Authors Homepage, Institutional Repository)
     * "Accepted Version (No Embargo, Locations: Funder Designated Location, etc.)
     * I'm open to other ideas here, but I like how Sherpa does some differentiation with icons, eg. https://v2.sherpa.ac.uk/id/publication/11114

Screenshot:
sherpa

I also find the display is odd if there no matching entry...here's what you see.  I'd expect it to just show a message saying "No Sherpa entries found" or similar.  The refresh button can remain though.
noSherpa

Beyond that, the basics work. It's just a bit of a confusing user experience.

File: src/app/submission/sections/sherpa-policies/content-accordion/content-accordion.component.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
+import { Component, Input } from '@angular/core';
+
+import { PermittedVersions } from '../../../../core/submission/models/sherpa-policies-details.model';
+
+@Component({
+  selector: 'ds-content-accordion',
+  templateUrl: './content-accordion.component.html',
+  styleUrls: ['./content-accordion.component.scss']
+})
+export class ContentAccordionComponent {
```
Comment (tdonohue) [inline src/app/submission/sections/sherpa-policies/content-accordion/content-accordion.component.ts:10]: Please add TypeDocs

File: src/app/submission/sections/sherpa-policies/metadata-information/metadata-information.component.ts [hunk 4]
```diff
@@ -0,0 +1,16 @@
+import { Component, Input } from '@angular/core';
+
+import { Metadata } from '../../../../core/submission/models/sherpa-policies-details.model';
+
+@Component({
+  selector: 'ds-metadata-information',
+  templateUrl: './metadata-information.component.html',
+  styleUrls: ['./metadata-information.component.scss']
+})
+export class MetadataInformationComponent {
```
Comment (tdonohue) [inline src/app/submission/sections/sherpa-policies/metadata-information/metadata-information.component.ts:10]: Please add TypeDocs to describe

File: src/app/submission/sections/sherpa-policies/publisher-policy/publisher-policy.component.ts [hunk 5]
```diff
@@ -0,0 +1,17 @@
+import { Component, Input } from '@angular/core';
+
+import { Policy } from '../../../../core/submission/models/sherpa-policies-details.model';
+
+@Component({
+  selector: 'ds-publisher-policy',
+  templateUrl: './publisher-policy.component.html',
+  styleUrls: ['./publisher-policy.component.scss']
+})
+export class PublisherPolicyComponent {
```
Comment (tdonohue) [inline src/app/submission/sections/sherpa-policies/publisher-policy/publisher-policy.component.ts:10]: Please add TypeDocs to describe

File: src/assets/i18n/en.json5 [hunk 6]
```diff
@@ -3989,6 +3993,60 @@
...
+
+  "submission.sections.sherpa.record.information.uri": "URI",
+
+
+
+
+
+
+
+
```
Comment (tdonohue) [inline src/assets/i18n/en.json5:4049]: Small thing but there's 7 blank lines here. We can remove most of these.

**Review by tdonohue — CHANGES_REQUESTED** [review]
@atarix83 : Code reviewed this quickly today (yesterday I only tested it).  Please see my notes inline..it's mostly requests for TypeDocs, but I also agree with the feedback from Art's review.

**Review by artlowel — COMMENTED** [review]
I've gotten a chance to test the feature as well now (last time I only did a code review), and I agree with @tdonohue's feedback that it's quite confusing

The panel looks just like any other submission panel, it even has a checkmark at the top. but it only provides info. The panel doesn't appear immediately after adding the ISSN, but only after the next save, so users may not link the two.

I think it would make more sense to either add it to the upload section, more similar to the way it worked in dspace 6:
screenshot 

Or perhaps as a notification after the item saves and a hit from sherpa has come back: "We found the publisher on Sherpa Romeo, click [here](#) to see their open access policies", which would then show it in an overlay or something.

If you want to keep the extra effort to a minimum, then i'd at least style that panel differently (e.g. in the bootstrap info style) and give it a lot more context. E.g. give it a better title, and perhaps add a paragraph explaining what sherpa romeo is, that it's just information to help you decide what to submit and that it doesn't automatically do anything to the item being submitted.

**Review by artlowel — CHANGES_REQUESTED** [review]
Currently there's a bunch of error when you start a new submission and there's no ISSN yet, because [sherpaData](https://github.com/4Science/dspace-angular/blob/d1f8bb6d7b04035729875c5053de47ea0904fbf2/src/app/submission/sections/sherpa-policies/section-sherpa-policies.component.html#L62) is undefined

Also I still don't think it's clear enough that it's a section that only provides info, and doesn't require you to fill anything in. You know the sectionType [here](https://github.com/4Science/dspace-angular/blob/166f4c3b25f5c0756d64df3204d999c5c1e90781/src/app/submission/sections/container/section-container.component.html#L3). When it's `sherpaPolicy` you could use a custom [NgbPanelHeader](https://ng-bootstrap.github.io/#/components/accordion/api#NgbPanelHeader) with e.g. a background in the info color, to clearly indicate that this section is different

**Comment by artlowel** [comment]
>>Currently there's a bunch of error when you start a new submission and there's no ISSN yet, because [sherpaData](https://github.com/4Science/dspace-angular/blob/d1f8bb6d7b04035729875c5053de47ea0904fbf2/src/app/submission/sections/sherpa-policies/section-sherpa-policies.component.html#L62) is undefined
>
>the bug is fixed

Thanks

>>Also I still don't think it's clear enough that it's a section that only provides info, and doesn't require you to fill anything in. You know the sectionType [here](https://github.com/4Science/dspace-angular/blob/166f4c3b25f5c0756d64df3204d999c5c1e90781/src/app/submission/sections/container/section-container.component.html#L3). When it's sherpaPolicy you could use a custom [NgbPanelHeader](https://ng-bootstrap.github.io/#/components/accordion/api#NgbPanelHeader) with e.g. a background in the info color, to clearly indicate that this section is different
>
>this is how it'd be following your suggestion
schreenshot

I mentioned that as an example. This is not a binary choice. If this is what happens if you give it the bootstrap `bg-info` class, then I agree, it's not very appealing. But that doesn't mean we should give up and fall back to just using the small icon in that case, because we agreed that wasn't very clear. It just means it needs some work. e.g. use a [card border](https://getbootstrap.com/docs/4.6/components/card/#border) instead, or use `text-white` on the header so you can read the text and the icon pops out more clearly etc.

**Review by tdonohue — CHANGES_REQUESTED** [review]
@atarix83 : I gave this a test today, and overall it's looking good to me.  However, I have a few small usability suggestions which I hope will address what @artlowel also pointed out... basically, I agree that it'd be nice to try to ensure the purpose of this new panel isn't misunderstood (since it's informational only).

It's easiest to describe my suggestions via screenshots.  Here's a screenshot of what I see when a an ISSN lookup in Sherpa is successful (I've added some suggestions into this screenshot which I describe below)
Sherpa-info-found-suggestion

Suggested minor changes:
1.  I'd recommend changing the heading of the panel to be "Publisher open access policy information"  This makes it clear what the purpose of the panel is
2. I'd recommend the heading section be stylized in *light blue*, it could be the same color as the info section that I circled in the screenshot.  This would make the heading appear more like an "info box", even when collapsed.
3. I'd recommend adding a paragraph description at the top of this panel, in order to describe where this data is gathered from & what you should do with it.  See my suggested text of "The below information..." in the screenshot above.
4. I'd recommend the circled text have the blue background removed...or change it to be a gray color.  I think the blue should be moved up to the heading, rather then midway through the panel.

The rest of this panel **looks great** to me.  I'm just trying to help improve the usability slightly by making it clearer to users that this is an informational panel.

Secondly, I'd recommend minor updates to the look of the panel when it first loads.  Again, here's a screenshot of how it currently looks:
Sherpa-no-info

1. Obviously, the same changes to the heading (as above) should be applied here.
2. Change the "No information available" text to no longer have the blue background. It can just be general text (as the blue background will move to the heading).  I'd recommend rewording this text to say: "No publisher policy information available. If your work has an associated ISSN, please enter it above to see any related publisher open access policies."

Beyond that, all my prior feedback has been addressed.  If any of these style changes are complex/difficult, I'm OK with delaying them as needed...but I feel they would improve the usability of this panel by clarifying that it's informational in nature.

**Comment by atarix83** [comment]
@artlowel @tdonohue 

we should have address most of your feedback, here the final result 

Schermata da 2022-06-06 15-58-16


@tdonohue 

you have suggested to not use the alert box for display section info message. But this is the way we've used also for other section, so to be consist we have left the alert box.
e.g. the upload step have a similar behavior for additional section information
Schermata da 2022-06-06 15-37-51

**Review by tdonohue — APPROVED** [review]
👍 Thanks @atarix83 !  I think the updated design looks much better & makes it clearer that this panel is informational.  So, I approve of this PR...  as users begin to use this feature more, we can always make additional usability improvements as necessary in future PRs.

That said, I noticed **one comment** from @artlowel that may not have been addressed.  It's a minor thing, but it would be worth cleaning up now if possible (see inline comment below).  So, I'd recommend we fix this before merging as well, if possible.

@artlowel : When you have a chance to look at this again, I believe your other feedback has all been addressed.

File: src/app/shared/alert/alert.component.ts [hunk 7]
```diff
@@ -33,7 +33,7 @@ export class AlertComponent {
   /**
    * The alert type
    */
-  @Input() type: AlertType;
+  @Input() type: AlertType|string;
```
[the remark this answers is not in the document]
Comment (atarix83) [reply src/app/shared/alert/alert.component.ts]: done

**Review by artlowel — APPROVED** [review]
Thanks @atarix83!

**Comment by tdonohue** [comment]
Merging as this is at +2, and the backend PR was just merged.  Thanks again @atarix83 and team!

**Pull request merged**